File: frontend/src/types/task.ts
export type TaskStatus = "backlog" | "new" | "in_progress" | "review" | "done" | "cancelled";
export type TaskPriority = "urgent" | "high" | "medium" | "low";
export type Visibility = "family" | "private";
export type UpdateType = "progress" | "status_change" | "comment" | "blocker";

export interface ChecklistItem {
  id: string;
  text: string;
  completed: boolean;
}

export interface UserBrief {
  id: number;
  display_name: string;
  email: string;
}

export interface Task {
  id: number;
  user_id: number;
  created_by_id: number;
  assignee_id: number | null;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  checklist: ChecklistItem[];
  source: string;
  visibility: Visibility;
  deadline: string | null;
  reminder_at: string | null;
  completed_at: string | null;
  kanban_order: number;
  created_at: string;
  updated_at: string;
  creator?: UserBrief;
  assignee?: UserBrief;
  owner_name?: string;
  tags: import("./tag").TagBrief[];
}

export interface TaskUpdateItem {
  id: number;
  task_id: number;
  author_id: number;
  type: UpdateType;
  content: string | null;
  old_status: string | null;
  new_status: string | null;
  progress_percent: number | null;
  created_at: string;
  author?: UserBrief;
}

export interface KanbanBoard {
  backlog: Task[];
  new: Task[];
  in_progress: Task[];
  review: Task[];
  done: Task[];
  cancelled: Task[];
}

export interface CreateTaskInput {
  title: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  deadline?: string;
  reminder_at?: string;
  checklist?: ChecklistItem[];
  assignee_id?: number;
  visibility?: Visibility;
  tag_ids?: number[];
}

export interface UpdateTaskInput {
  title?: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  deadline?: string | null;
  reminder_at?: string | null;
  checklist?: ChecklistItem[];
  assignee_id?: number | null;
  visibility?: Visibility;
  tag_ids?: number[];
}

export interface CreateUpdateInput {
  type: UpdateType;
  content?: string;
  progress_percent?: number;
}

export interface TaskFilters {
  status?: string;
  priority?: string;
  assignee_id?: number;
  search?: string;
  tag_id?: number;
}

export const TASK_STATUS_LABELS: Record<TaskStatus, string> = {
  backlog: "Backlog",
  new: "New",
  in_progress: "In Progress",
  review: "Review",
  done: "Done",
  cancelled: "Cancelled",
};

export const TASK_STATUS_ORDER: TaskStatus[] = [
  "backlog",
  "new",
  "in_progress",
  "review",
  "done",
  "cancelled",
];

export const PRIORITY_COLORS: Record<TaskPriority, string> = {
  urgent: "text-destructive",
  high: "text-accent-amber",
  medium: "text-accent-foreground",
  low: "text-tertiary",
};

export const PRIORITY_BG_COLORS: Record<TaskPriority, string> = {
  urgent: "bg-destructive-muted text-destructive",
  high: "bg-accent-amber-muted text-accent-amber",
  medium: "bg-accent-muted text-accent-foreground",
  low: "bg-surface-hover text-tertiary",
};

export const PRIORITY_DOT_COLORS: Record<TaskPriority, string> = {
  urgent: "bg-destructive",
  high: "bg-accent-amber",
  medium: "bg-accent-foreground",
  low: "bg-tertiary",
};

export const PRIORITY_BORDER_COLORS: Record<TaskPriority, string> = {
  urgent: "border-l-destructive",
  high: "border-l-[var(--accent-amber)]",
  medium: "border-l-primary",
  low: "border-l-[var(--text-tertiary)]",
};

export const PRIORITY_BORDER_CSS_VARS: Record<TaskPriority, string> = {
  urgent: "var(--destructive)",
  high: "var(--accent-amber)",
  medium: "var(--primary)",
  low: "var(--text-tertiary)",
};

export const PRIORITY_LABELS: Record<TaskPriority, string> = {
  urgent: "Urgent",
  high: "High",
  medium: "Medium",
  low: "Low",
};

export const DEFAULT_HIDDEN_COLUMNS: TaskStatus[] = ["review", "cancelled"];

File: backend/app/schemas/task.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel

from app.models.task import TaskPriority, TaskSource, TaskStatus, UpdateType, Visibility
from app.schemas.tag import TagBrief


class ChecklistItem(BaseModel):
    id: str
    text: str
    completed: bool = False


class UserBrief(BaseModel):
    id: int
    display_name: str
    email: str

    model_config = {"from_attributes": True}


# ── Task schemas ─────────────────────────────────────────────────────────────


class TaskCreate(BaseModel):
    title: str
    description: Optional[str] = None
    status: TaskStatus = TaskStatus.new
    priority: TaskPriority = TaskPriority.medium
    deadline: Optional[datetime] = None
    reminder_at: Optional[datetime] = None
    reminder_floating: bool = False
    checklist: list[ChecklistItem] = []
    assignee_id: Optional[int] = None
    visibility: Visibility = Visibility.family
    tag_ids: list[int] = []


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[TaskStatus] = None
    priority: Optional[TaskPriority] = None
    deadline: Optional[datetime] = None
    reminder_at: Optional[datetime] = None
    reminder_floating: Optional[bool] = None
    checklist: Optional[list[ChecklistItem]] = None
    assignee_id: Optional[int] = None
    visibility: Optional[Visibility] = None
    tag_ids: Optional[list[int]] = None


class LinkedEventBrief(BaseModel):
    id: int
    title: str
    start_time: datetime
    end_time: datetime

    model_config = {"from_attributes": True}


class TaskReorder(BaseModel):
    task_id: int
    after_task_id: Optional[int] = None
    before_task_id: Optional[int] = None


class TaskResponse(BaseModel):
    id: int
    user_id: int
    created_by_id: int
    assignee_id: Optional[int]
    title: str
    description: Optional[str]
    status: TaskStatus
    priority: TaskPriority
    checklist: list[dict]
    source: TaskSource
    visibility: Visibility
    deadline: Optional[datetime]
    reminder_at: Optional[datetime]
    reminder_floating: bool
    reminder_dismissed: bool
    completed_at: Optional[datetime]
    kanban_order: float = 0
    created_at: datetime
    updated_at: datetime

    creator: Optional[UserBrief] = None
    assignee: Optional[UserBrief] = None
    owner_name: Optional[str] = None
    linked_events: list[LinkedEventBrief] = []
    tags: list[TagBrief] = []

    model_config = {"from_attributes": True}


# ── TaskUpdate (timeline) schemas ────────────────────────────────────────────


class TaskUpdateCreate(BaseModel):
    type: UpdateType
    content: Optional[str] = None
    progress_percent: Optional[int] = None


class TaskUpdateResponse(BaseModel):
    id: int
    task_id: int
    author_id: int
    type: UpdateType
    content: Optional[str]
    old_status: Optional[str]
    new_status: Optional[str]
    progress_percent: Optional[int]
    created_at: datetime

    author: Optional[UserBrief] = None

    model_config = {"from_attributes": True}


# ── Kanban schema ─────────────────────────────────────────────────────────────


class KanbanBoard(BaseModel):
    backlog: list[TaskResponse] = []
    new: list[TaskResponse] = []
    in_progress: list[TaskResponse] = []
    review: list[TaskResponse] = []
    done: list[TaskResponse] = []
    cancelled: list[TaskResponse] = []
